In [1]:
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Intro to Generating and Executing Python Code with Gemini 2.5 Flash

<table align="left">
  <td style="text-align: center">
    <a href="https://colab.research.google.com/github/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb">
      <img width="32px" src="https://www.gstatic.com/pantheon/images/bigquery/welcome_page/colab-logo.svg" alt="Google Colaboratory logo"><br> Open in Colab
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/colab/import/https:%2F%2Fraw.githubusercontent.com%2FGoogleCloudPlatform%2Fgenerative-ai%2Fmain%2Fgemini%2Fcode-execution%2Fintro_code_execution.ipynb">
      <img width="32px" src="https://lh3.googleusercontent.com/JmcxdQi-qOpctIvWKgPtrzZdJJK-J3sWE1RsfjZNwshCFgE_9fULcNpuXYTilIR2hjwN" alt="Google Cloud Colab Enterprise logo"><br> Open in Colab Enterprise
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/workbench/deploy-notebook?download_url=https://raw.githubusercontent.com/GoogleCloudPlatform/generative-ai/main/gemini/code-execution/intro_code_execution.ipynb">
      <img src="https://www.gstatic.com/images/branding/gcpiconscolors/vertexai/v1/32px.svg" alt="Vertex AI logo"><br> Open in Vertex AI Workbench
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb">
      <img width="32px" src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub logo"><br> View on GitHub
    </a>
  </td>
</table>

<div style="clear: both;"></div>

<b>Share to:</b>

<a href="https://www.linkedin.com/sharing/share-offsite/?url=https%3A//github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb" target="_blank">
  <img width="20px" src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="LinkedIn logo">
</a>

<a href="https://bsky.app/intent/compose?text=https%3A//github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb" target="_blank">
  <img width="20px" src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="Bluesky logo">
</a>

<a href="https://twitter.com/intent/tweet?url=https%3A//github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb" target="_blank">
  <img width="20px" src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X logo">
</a>

<a href="https://reddit.com/submit?url=https%3A//github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb" target="_blank">
  <img width="20px" src="https://redditinc.com/hubfs/Reddit%20Inc/Brand/Reddit_Logo.png" alt="Reddit logo">
</a>

<a href="https://www.facebook.com/sharer/sharer.php?u=https%3A//github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb" target="_blank">
  <img width="20px" src="https://upload.wikimedia.org/wikipedia/commons/5/51/Facebook_f_logo_%282019%29.svg" alt="Facebook logo">
</a>

| | |
|-|-|
| Author(s) |  [Kristopher Overholt](https://github.com/koverholt/) |

## Overview

This notebook introduces the code execution capabilities of the [Gemini 2.5 Flash model](https://cloud.google.com/vertex-ai/generative-ai/docs/models), a new multimodal generative AI model from Google [DeepMind](https://deepmind.google/). Gemini 2.5 Flash offers improvements in speed, quality, and advanced reasoning capabilities including enhanced understanding, coding, and instruction following.

## Code Execution

A key feature of this model is [code execution](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/code-execution), which is the ability to generate and execute Python code directly within the API. If you want the API to generate and run Python code and return the results, you can use code execution as demonstrated in this notebook.

This code execution capability enables the model to generate code, execute and observe the results, correct the code if needed, and learn iteratively from the results until it produces a final output. This is particularly useful for applications that involve code-based reasoning such as solving mathematical equations or processing text.

## Objectives

In this tutorial, you will learn how to generate and execute code using the Gemini API in Vertex AI and the Google Gen AI SDK for Python with the Gemini 2.5 Flash model.

You will complete the following tasks:

- Generating and running sample Python code from text prompts
- Exploring data using code execution in multi-turn chats
- Using code execution in streaming sessions

## Getting started

### Install Google Gen AI SDK for Python


In [1]:
%pip install --upgrade --quiet google-genai

Note: you may need to restart the kernel to use updated packages.


### Import libraries

In [2]:
import os

from IPython.display import Markdown, display
from google import genai
from google.genai.types import GenerateContentConfig, Tool, ToolCodeExecution

### Connect to a generative AI API service

Google Gen AI APIs and models including Gemini are available in the following two API services:

- [Google AI for Developers](https://ai.google.dev/gemini-api/docs): Experiment, prototype, and deploy small projects.
- [Vertex AI](https://cloud.google.com/vertex-ai/generative-ai/docs): Build enterprise-ready projects on Google Cloud.
The Google Gen AI SDK provides a unified interface to these two API services.

This notebook shows how to use the Google Gen AI SDK with the Gemini API in Vertex AI.

### Set Google Cloud project information and create client

To get started using Vertex AI, you must have an existing Google Cloud project and [enable the Vertex AI API](https://console.cloud.google.com/flows/enableapi?apiid=aiplatform.googleapis.com).

Learn more about [setting up a project and a development environment](https://cloud.google.com/vertex-ai/docs/start/cloud-environment).

In [3]:
import os

PROJECT_ID = "qwiklabs-gcp-04-135368faf912"
LOCATION = os.environ.get("GOOGLE_CLOUD_REGION", "global")
client = genai.Client(vertexai=True, project=PROJECT_ID, location=LOCATION)


## Working with code execution in Gemini 2.5 Flash

### Load the Gemini model

The following code loads the Gemini 2.5 Flash model. You can learn about all Gemini models on Vertex AI by visiting the [documentation](https://cloud.google.com/vertex-ai/generative-ai/docs/learn/models):

In [4]:
MODEL_ID = "gemini-2.5-flash"  # @param {type: "string"}

### Define the code execution tool

The following code initializes the code execution tool by passing `code_execution` in a `Tool` definition.

Later we'll register this tool with the model that it can use to generate and run Python code:

In [5]:
code_execution_tool = Tool(code_execution=ToolCodeExecution())

### Generate and execute code

The following code sends a prompt to the Gemini model, asking it to generate and execute Python code to calculate the sum of the first 50 prime numbers. The code execution tool is passed in so the model can generate and run the code:

In [6]:
PROMPT = """What is the sum of the first 50 prime numbers?
Generate and run code for the calculation."""

response = client.models.generate_content(
    model=MODEL_ID,
    contents=PROMPT,
    config=GenerateContentConfig(
        tools=[code_execution_tool],
        temperature=0,
    ),
)

### View the generated code

The following code iterates through the response and displays any generated Python code by checking for `part.executable_code` in the response parts:

In [7]:
for part in response.candidates[0].content.parts:
    if part.executable_code:
        display(
            Markdown(
                f"""
```py
{part.executable_code.code}
```
"""
            )
        )


```py
import math

def is_prime(n):
    if n < 2:
        return False
    for i in range(2, int(math.sqrt(n)) + 1):
        if n % i == 0:
            return False
    return True

prime_numbers = []
num = 2
while len(prime_numbers) < 50:
    if is_prime(num):
        prime_numbers.append(num)
    num += 1

sum_of_primes = sum(prime_numbers)

print(f"The first 50 prime numbers are: {prime_numbers}")
print(f"The sum of the first 50 prime numbers is: {sum_of_primes}")
```


### View the code execution results

The following code iterates through the response and displays the execution result and outcome by checking for `part.code_execution_result` in the response parts:

In [8]:
for part in response.candidates[0].content.parts:
    if part.code_execution_result:
        display(Markdown(f"`{part.code_execution_result.output}`"))
        print("\nOutcome:", part.code_execution_result.outcome)

`The first 50 prime numbers are: [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229]
The sum of the first 50 prime numbers is: 5117
`


Outcome: Outcome.OUTCOME_OK


Great! Now you have the answer (`5117`) as well as the generated (and verified via execution!) Python code.

At this point in your application, you would save the output code, result, or outcome and display it to the end-user or use it downstream in your application.

### Code execution in a chat session

This section shows how to use code execution in an interactive chat with history using the Gemini API.

You can use `client.chats.create` to create a chat session and passes in the code execution tool, enabling the model to generate and run code:

In [9]:
chat = client.chats.create(
    model=MODEL_ID,
    config=GenerateContentConfig(
        tools=[code_execution_tool],
        temperature=0,
    ),
)

You'll start the chat by asking the model to generate sample time series data with noise and then output a sample of 10 data points:

In [10]:
PROMPT = """Create sample time series data of temperature vs. time in a test furnace.
Add noise to the data. Output a sample of 10 data points from the time series data."""

response = chat.send_message(PROMPT)

Now you can iterate through the response to display any generated Python code and execution results by checking for `part.executable_code` and `part.code_execution_result` in the response parts:

In [11]:
for part in response.candidates[0].content.parts:
    if part.executable_code:
        display(
            Markdown(
                f"""
```py
{part.executable_code.code}
```
"""
            )
        )
    if part.code_execution_result:
        display(Markdown(f"`{part.code_execution_result.output}`"))
        print("\nOutcome:", part.code_execution_result.outcome)


```py
import numpy as np

# 1. Define time range
# Let's simulate data for 0 to 60 minutes, with 1-minute intervals
time_points = np.arange(0, 60, 1) # 60 data points

# 2. Create a base temperature profile
initial_temp = 25  # Starting temperature in Celsius
target_temp = 1000 # Target stable temperature in Celsius
heating_rate_constant = 0.1 # Controls how fast the temperature rises

# Using an exponential rise model: T(t) = T_initial + (T_target - T_initial) * (1 - exp(-k*t))
base_temperature = initial_temp + (target_temp - initial_temp) * (1 - np.exp(-heating_rate_constant * time_points))

# 3. Add random noise
# Standard deviation for the noise (e.g., +/- 5 degrees)
noise_std_dev = 7
noise = np.random.normal(0, noise_std_dev, len(time_points))
noisy_temperature = base_temperature + noise

# 4. Select a sample of 10 data points
# For demonstration, let's pick 10 points from the beginning of the series
sample_indices = np.arange(0, 10) # First 10 points
# Alternatively, to pick random points:
# sample_indices = np.random.choice(len(time_points), 10, replace=False)

sample_time = time_points[sample_indices]
sample_temperature = noisy_temperature[sample_indices]

# Output the sample data
print("Sample Time Series Data (Temperature vs. Time):")
print("------------------------------------------------")
print("Time (minutes) | Temperature (Celsius)")
print("------------------------------------------------")
for t, temp in zip(sample_time, sample_temperature):
    print(f"{t:<14.2f} | {temp:<20.2f}")

```


`Sample Time Series Data (Temperature vs. Time):
------------------------------------------------
Time (minutes) | Temperature (Celsius)
------------------------------------------------
0.00           | 10.90               
1.00           | 117.73              
2.00           | 205.21              
3.00           | 267.77              
4.00           | 334.28              
5.00           | 398.33              
6.00           | 470.56              
7.00           | 520.22              
8.00           | 565.54              
9.00           | 597.56              
`


Outcome: Outcome.OUTCOME_OK


Now you can ask the model to add a smoothed data series to the time series data:

In [12]:
PROMPT = "Now add a data series that smooths the sample data."

response = chat.send_message(PROMPT)

And then display the generated Python code and execution results:

In [13]:
for part in response.candidates[0].content.parts:
    if part.executable_code:
        display(
            Markdown(
                f"""
```py
{part.executable_code.code}
```
"""
            )
        )
    if part.code_execution_result:
        display(Markdown(f"`{part.code_execution_result.output}`"))
        print("\nOutcome:", part.code_execution_result.outcome)


```py
import numpy as np
import pandas as pd # pandas is excellent for moving averages

# Regenerate the original full time series data (as in the previous step)
time_points_full = np.arange(0, 60, 1)
initial_temp = 25
target_temp = 1000
heating_rate_constant = 0.1
base_temperature_full = initial_temp + (target_temp - initial_temp) * (1 - np.exp(-heating_rate_constant * time_points_full))
noise_std_dev = 7
# To ensure the exact same noise, we would need to set a random seed.
# For this demonstration, a new noise generation is fine, but the base profile is the same.
np.random.seed(42) # Setting a seed for reproducibility of noise
noise_full = np.random.normal(0, noise_std_dev, len(time_points_full))
noisy_temperature_full = base_temperature_full + noise_full

# Select the same 10 sample data points
sample_indices = np.arange(0, 10)
sample_time = time_points_full[sample_indices]
sample_temperature = noisy_temperature_full[sample_indices]

# Now, add the smoothing series
# Convert to a pandas Series for easy moving average calculation
temp_series = pd.Series(sample_temperature)

# Apply a moving average filter (e.g., window size of 3)
# .rolling(window=3) creates a rolling window object
# .mean() calculates the mean within each window
# .fillna(method='bfill') or .fillna(method='ffill') can handle NaNs at the start/end
# For simplicity, we'll just let the first few values be NaN, or fill them with the original value
smoothed_temperature = temp_series.rolling(window=3, min_periods=1, center=True).mean()

# Output the sample data with the smoothed series
print("Sample Time Series Data (Temperature vs. Time) with Smoothing:")
print("----------------------------------------------------------------")
print("Time (minutes) | Noisy Temp (C) | Smoothed Temp (C)")
print("----------------------------------------------------------------")
for t, noisy_t, smoothed_t in zip(sample_time, sample_temperature, smoothed_temperature):
    print(f"{t:<14.2f} | {noisy_t:<14.2f} | {smoothed_t:<15.2f}")

```


`Sample Time Series Data (Temperature vs. Time) with Smoothing:
----------------------------------------------------------------
Time (minutes) | Noisy Temp (C) | Smoothed Temp (C)
----------------------------------------------------------------
0.00           | 28.48          | 72.65          
1.00           | 116.82         | 117.19         
2.00           | 206.27         | 203.82         
3.00           | 288.36         | 279.81         
4.00           | 344.80         | 346.72         
5.00           | 406.99         | 409.25         
6.00           | 475.96         | 468.05         
7.00           | 521.20         | 518.59         
8.00           | 558.62         | 562.40         
9.00           | 607.39         | 583.01         
`


Outcome: Outcome.OUTCOME_OK


Finally, you can ask the model to generate descriptive statistics for the time series data:

In [14]:
PROMPT = "Now generate and output descriptive statistics on the time series data."

response = chat.send_message(PROMPT)

And then display the generated Python code and execution results:

In [15]:
for part in response.candidates[0].content.parts:
    if part.executable_code:
        display(
            Markdown(
                f"""
```py
{part.executable_code.code}
```
"""
            )
        )
    if part.code_execution_result:
        display(Markdown(f"`{part.code_execution_result.output}`"))
        print("\nOutcome:", part.code_execution_result.outcome)


```py
import numpy as np
import pandas as pd

# Regenerate the original full time series data (as in the previous steps)
time_points_full = np.arange(0, 60, 1)
initial_temp = 25
target_temp = 1000
heating_rate_constant = 0.1
base_temperature_full = initial_temp + (target_temp - initial_temp) * (1 - np.exp(-heating_rate_constant * time_points_full))
noise_std_dev = 7
np.random.seed(42) # Setting a seed for reproducibility of noise
noise_full = np.random.normal(0, noise_std_dev, len(time_points_full))
noisy_temperature_full = base_temperature_full + noise_full

# Select the same 10 sample data points
sample_indices = np.arange(0, 10)
sample_time = time_points_full[sample_indices]
sample_temperature = noisy_temperature_full[sample_indices]

# Apply smoothing
temp_series = pd.Series(sample_temperature)
smoothed_temperature = temp_series.rolling(window=3, min_periods=1, center=True).mean()

# Create a DataFrame for easier statistics generation
data_df = pd.DataFrame({
    'Time (minutes)': sample_time,
    'Noisy Temp (C)': sample_temperature,
    'Smoothed Temp (C)': smoothed_temperature
})

# Generate descriptive statistics for the temperature columns
print("Descriptive Statistics for Noisy Temperature:")
print(data_df['Noisy Temp (C)'].describe())
print("\nDescriptive Statistics for Smoothed Temperature:")
print(data_df['Smoothed Temp (C)'].describe())

```


`Descriptive Statistics for Noisy Temperature:
count     10.000000
mean     355.489493
std      194.402766
min       28.476999
25%      226.794363
50%      375.896265
75%      509.891815
max      607.392502
Name: Noisy Temp (C), dtype: float64

Descriptive Statistics for Smoothed Temperature:
count     10.000000
mean     356.148895
std      183.282792
min       72.646333
25%      222.815417
50%      377.985275
75%      505.958794
max      583.005221
Name: Smoothed Temp (C), dtype: float64
`


Outcome: Outcome.OUTCOME_OK


This chat example demonstrates how you can use the Gemini API with code execution as a powerful tool for exploratory data analysis and more. Go forth and adapt this approach to your own projects and use cases!

### Code execution in a streaming session

You can also use the code execution functionality with streaming output from the Gemini API.

The following code demonstrates how the Gemini API can generate and execute code while streaming the results:

In [16]:
PROMPT = """Generate a list of 20 random names, then create a new list with just
the names containing the letter 'a', then output the number of names that
contain 'a' and finally show me that new list."""

for chunk in client.models.generate_content_stream(
    model=MODEL_ID,
    contents=PROMPT,
    config=GenerateContentConfig(
        tools=[code_execution_tool],
        temperature=0,
    ),
):
    if chunk.candidates and chunk.candidates[0].content:
        if chunk.candidates[0].content.parts is not None:
            for part in chunk.candidates[0].content.parts:
                if part.text:
                    display(Markdown("#### Natural language stream"))
                    display(Markdown(part.text))
                    display(Markdown("---"))
                if part.executable_code:
                    display(Markdown("#### Code stream"))
                    display(
                        Markdown(
                            f"""
    ```py
    {part.executable_code.code}
    ```
    """
                        )
                    )
                    display(Markdown("---"))
                if part.code_execution_result:
                    display(Markdown("#### Code result"))
                    display(
                        Markdown(
                            f"""
    ```
    {part.code_execution_result.output}
    ```
    """
                        )
                    )
                    display(Markdown("---"))

#### Code stream


    ```py
    import random
import string

def generate_random_name(min_len=4, max_len=8):
    """Generates a random string of lowercase letters to serve as a 'name'."""
    length = random.randint(min_len, max_len)
    # Ensure a mix of characters, including 'a' to make the filtering interesting
    characters = string.ascii_lowercase
    name = ''.join(random.choice(characters) for _ in range(length))
    return name

# 1. Generate a list of 20 random names
random_names = [generate_random_name() for _ in range(20)]
print(f"Original list of 20 random names:\n{random_names}\n")

# 2. Create a new list with just the names containing the letter 'a'
names_with_a = [name for name in random_names if 'a' in name.lower()]

# 3. Output the number of names that contain 'a'
number_of_names_with_a = len(names_with_a)
print(f"Number of names containing the letter 'a': {number_of_names_with_a}\n")

# 4. Show the new list
print(f"List of names containing the letter 'a':\n{names_with_a}")
    ```
    

---

#### Code result


    ```
    Original list of 20 random names:
['mchyqsbv', 'obbory', 'kgzu', 'vvbe', 'texjaghi', 'imtd', 'iwcxu', 'dggswyj', 'vyumv', 'pwmwyhb', 'cxvreqsq', 'wtwivwnq', 'ijpj', 'xrcee', 'xomdq', 'dgzzd', 'mddyly', 'jrskocf', 'mqbww', 'wzxuryew']

Number of names containing the letter 'a': 1

List of names containing the letter 'a':
['texjaghi']

    ```
    

---

#### Natural language stream

Here's the process:

First, I generated a list of 20 random names (random strings of lowercase letters for this exercise).

Then, I filtered this list to create a new list containing only the names that include the letter 'a' (

---

#### Natural language stream

case-insensitive).

Finally, I counted the names in this new list and displayed both the count and the list itself.



---

#### Code stream


    ```py
    import random
import string

def generate_random_name(min_len=4, max_len=8):
    """Generates a random string of lowercase letters to serve as a 'name'."""
    length = random.randint(min_len, max_len)
    characters = string.ascii_lowercase
    name = ''.join(random.choice(characters) for _ in range(length))
    return name

# 1. Generate a list of 20 random names
random_names = [generate_random_name() for _ in range(20)]
print(f"Original list of 20 random names:\n{random_names}\n")

# 2. Create a new list with just the names containing the letter 'a'
names_with_a = [name for name in random_names if 'a' in name.lower()]

# 3. Output the number of names that contain 'a'
number_of_names_with_a = len(names_with_a)
print(f"Number of names containing the letter 'a': {number_of_names_with_a}\n")

# 4. Show the new list
print(f"List of names containing the letter 'a':\n{names_with_a}")
    ```
    

---

#### Code result


    ```
    Original list of 20 random names:
['gketet', 'epacdu', 'urgz', 'qkrqx', 'plnreiqi', 'zciz', 'qeyvv', 'xwdl', 'yfsxc', 'jnnzxqdp', 'cskfaawn', 'szvcmc', 'otbvr', 'yxccsu', 'ebrgxe', 'kpwqloxc', 'ghju', 'smvcnzm', 'ejhkz', 'weoj']

Number of names containing the letter 'a': 2

List of names containing the letter 'a':
['epacdu', 'cskfaawn']

    ```
    

---

#### Natural language stream

Here

---

#### Natural language stream

 are the results:

**Original list of 20 random names:**


---

#### Natural language stream

`['gketet', 'epacdu', 'urgz', 'qkrqx

---

#### Natural language stream

', 'plnreiqi', 'zciz', 'qeyvv', 'xwdl', 'yfsxc', 'jnnzxqdp', 'csk

---

#### Natural language stream

faawn', 'szvcmc', 'otbvr', 'yxccsu', 'ebrgxe', 'kpwqloxc', 'ghju',

---

#### Natural language stream

 'smvcnzm', 'ejhkz', 'weoj']`

**Number of names containing the letter 'a':** `2`

**List of names containing the letter 'a':**
`['epacdu', 'csk

---

#### Natural language stream

faawn']`

---

This streaming example demonstrated how the Gemini API can generate, execute code, and provide results within a streaming session.

## Summary

Refer to the [documentation](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/code-execution) for more details about code execution, and in particular, the [recommendations](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/code-execution#code-execution-vs-function-calling) regarding differences between code execution and [function calling](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/function-calling).

### Next steps

- See the [Google Gen AI SDK reference docs](https://googleapis.github.io/python-genai/)
- Explore other notebooks in the [Google Cloud Generative AI GitHub repository](https://github.com/GoogleCloudPlatform/generative-ai)
- Explore AI models in [Model Garden](https://cloud.google.com/vertex-ai/generative-ai/docs/model-garden/explore-models)